x-pack/plugins/security_solution/server/lib/detection_engine/signals/search_after_bulk_create.ts

```diff
@@ -92,23 +97,86 @@ export const searchAfterAndBulkCreate = async ({
+          const singleSearchAfterDefaultTimestamp = singleSearchAfter({
+            buildRuleMessage,
+            searchAfterSortId: backupSortId,
+            index: inputIndexPattern,
+            from: tuple.from.toISOString(),
+            to: tuple.to.toISOString(),
+            services,
+            logger,
+            filter,
+            pageSize: tuple.maxSignals < pageSize ? Math.ceil(tuple.maxSignals) : pageSize, // maximum number of docs to receive per search result.
+            timestampOverride: ruleParams.timestampOverride,
+            excludeDocsWithTimestampOverride: true,
+          });
+          const {
+            searchResult: searchResultB,
+            searchDuration: searchDurationB,
+            searchErrors: searchErrorsB,
+          } = await singleSearchAfterDefaultTimestamp;
+
+          // call this function setSortIdOrExit()
+          const lastSortId = searchResultB?.hits?.hits[searchResultB.hits.hits.length - 1]?.sort;
+          if (lastSortId != null && lastSortId.length !== 0) {
+            backupSortId = lastSortId[0];
+          } else {
+            // if no sort id on backup search and the initial search result was also empty
+            logger.debug(buildRuleMessage('backupSortIds was empty on searchResultB'));
+            backupSortId = undefined;
+          }
+
+          mergedSearchResults = mergeSearchResults([mergedSearchResults, searchResultB]);
+
+          // merge the search result from the secondary search with the first
+          toReturn = mergeReturns([
+            toReturn,
+            createSearchAfterReturnTypeFromResponse({
+              searchResult: mergedSearchResults,
+              timestampOverride: undefined,
+            }),
+            createSearchAfterReturnType({
+              searchAfterTimes: [searchDurationB],
+              errors: searchErrorsB,
+            }),
+          ]);
+        }
+
+        const { searchResult, searchDuration, searchErrors } = await singleSearchAfterPromise;
+        mergedSearchResults = mergeSearchResults([mergedSearchResults, searchResult]);
         toReturn = mergeReturns([
           toReturn,
           createSearchAfterReturnTypeFromResponse({
-            searchResult,
+            searchResult: mergedSearchResults,
             timestampOverride: ruleParams.timestampOverride,
           }),
           createSearchAfterReturnType({
             searchAfterTimes: [searchDuration],
             errors: searchErrors,
           }),
         ]);
+
+        // we are guaranteed to have searchResult hits at this point
+        // because we check before if the totalHits or
+        // searchResult.hits.hits.length is 0
+        // call this function setSortIdOrExit()
+        const lastSortId = searchResult.hits.hits[searchResult.hits.hits.length - 1]?.sort;
+        if (lastSortId != null && lastSortId.length !== 0) {
+          sortId = lastSortId[0];
+        } else {
+          sortId = undefined;
```

Comment:
Setting `sortId` to `undefined` will cause the next search using `sortId` to start from the beginning of the result set again. We should use a boolean flag here instead to indicate that the queries with `sortId` are finished and on future iterations of this while loop we can skip those queries.

With the current implementation if we have a M pages of results for the primary query and N pages for the backup query then the first time `sortId` and `backupSortId` are both undefined (and thus the loop can exit) is after LeastCommonMultiple(M, N) iterations of the loop rather than Max(M, N). E.G. if we have 6 and 7 pages, respectively, then the while loop will take 42 iterations to terminate rather than only 7 iterations.

Reply:
Good catch! I just pushed an update with this change.